ucp/endpoint_reuse.py

```diff
@@ -0,0 +1,125 @@
+import ctypes
+import pickle
+import uuid
+
+from . import core
+
+
+class EPHandle:
+    def __init__(self, ep):
+        self.ep = ep
+        self.refcount = 1
+
+
+class EndpointReuse:
+    """Class to seamlessly reuse endpoints.
+
+    It uses the tag feature of send/recv to separate "virtual" endpoint
+    pairs from each other.
+
+    Warning
+    -------
+    When closing a reused endpoint, the peer might not be notified.
+
+    Performance
+    -----------
+    The overhead of creating endpoints is increased but the performance
+    of the created connections is the same.
+
+    Connection Protocol
+    -------------------
+    1) Client connect to server using a new endpoint.
+    2) Client send the IDs of all its existing endpoints.
+    3) Server receives the IDs from the client and checks if it has a
+       matching existing endpoint. It then sends the matching ID to the
+       client or zero if no match.
+    4) The client and server now continue with either the existing matching
+       endpoints or the new endpoints (which are registered for later reuse).
+    """
+
+    existing_endpoints = {}
+
+    def __init__(self, handle, tag):
+        self.handle = handle
+        self.tag = tag
+
+    @classmethod
+    async def create_endpoint(cls, ip, port):
+        tag = ctypes.c_uint32(uuid.uuid4().int).value
+        ep_new = await core.create_endpoint(ip, port)
+
+        existing_endpoints = list(cls.existing_endpoints.values())
+        my_ep_ids = []
+        for ep in existing_endpoints:
+            if not ep.ep.closed():
+                ep.refcount += 1
+                my_ep_ids.append(ep.ep._msg_tag_recv)
+
+        await ep_new.send_obj(pickle.dumps((my_ep_ids, tag)))
+        reuse_ep_id = pickle.loads(await ep_new.recv_obj())
+
+        for ep in existing_endpoints:
+            if not ep.ep.closed():
+                ep.refcount -= 1
+                if ep.refcount == 0:
+                    await ep.ep.close()
+
+        if reuse_ep_id:
+            reuse_ep = cls.existing_endpoints[reuse_ep_id]
+            reuse_ep.refcount += 1
+            await ep_new.close()
+        else:
+            reuse_ep = EPHandle(ep_new)
+            assert ep_new._msg_tag_send not in cls.existing_endpoints
+            cls.existing_endpoints[ep_new._msg_tag_send] = reuse_ep
+        return cls(reuse_ep, tag)
+
+    @classmethod
+    def create_listener(cls, cb_coroutine, port):
+        async def _handle(ep_new):
```

Comment:
Is it possible to implement this without a closure?

Reply:
Yes, I can move `_handle` outside `EndpointReuse` and use `functools.partial()` but again I am not sure it is worth the extra code?